raft: update LeadEpoch in response to MsgFortify

arulajmani opened the pull request:
This patch partially implements the handling of MsgFortify on peers.
In particular, peers (including the leader) now check whether they
support the leader when handling a fortification request. They persist
the leadEpoch from the StoreLiveness fabric if they do. For now, the
fortification request is trivially rejected regardless; this will change
shortly.

Most of this patch is building out some of the test infra we'll use to
set up the underlying StoreLiveness fabric.

Informs https://github.com/cockroachdb/cockroach/issues/125261

Release note: None

arulajmani committed:
raft: include LeadEpoch when printing out HardState

This was missed when addressing #124497.

Epic: none

Release note: None

arulajmani committed:
raft: update LeadEpoch in response to MsgFortify

This patch partially implements the handling of MsgFortify on peers.
In particular, peers (including the leader) now check whether they
support the leader when handling a fortification request. They persist
the leadEpoch from the StoreLiveness fabric if they do. For now, the
fortification request is trivially rejected regardless; this will change
shortly.

Most of this patch is building out some of the test infra we'll use to
set up the underlying StoreLiveness fabric.

Informs https://github.com/cockroachdb/cockroach/issues/125261

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/128845)
<!-- Reviewable:end -->

nvb reviewed (commented):
> Reviewed 26 of 26 files at r1, 36 of 36 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128845)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/kv/kvserver/replica_store_liveness.go` line 48 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48eOsaAXKmdbz011zG:-O48eOsaAXKmdbz011zH:b-a5lh4k) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/kv/kvserver/replica_store_liveness.go#L48)):*
> ```Go
> 		return 0, false
> 	}
> 	if r.store.storeLiveness == nil {
> ```

Do we want to keep this indefinitely, or add a TODO to remove it once `storeLiveness` is configured in the `Store` constructor?
___
*[`pkg/raft/raft.go` line 521 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48f547CJIxVhrtL8MT:-O48f547CJIxVhrtL8MU:bgeupt5) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/raft.go#L521)):*
> ```Go
> 		}
> 	}
> 	if m.Type == pb.MsgAppResp || m.Type == pb.MsgVoteResp || m.Type == pb.MsgPreVoteResp {
> ```

Reminder that we're going to need to add `MsgFortifyResp` to this list. We may want to do so now, to avoid the data-driven churn.
___
*[`pkg/raft/raft.go` line 1713 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48fH1JDlqy6qJNAk4C:-O48fH1JDlqy6qJNAk4D:b5gkggh) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/raft.go#L1713)):*
> ```Go
> 		r.handleSnapshot(m)
> 	case pb.MsgFortifyLeader:
> 		r.lead = m.From
> ```

We haven't quite decided how we want to use the `r.electionElapsed` field with fortification, but we may end up wanting to reset it to 0 here, for parity with the other cases.
___
*[`pkg/raft/raft.go` line 1881 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48en2g9rBPkM8st_1F:-O48en2g9rBPkM8st_1G:bvlx7qt) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/raft.go#L1881)):*
> ```Go
> 
> func (r *raft) handleFortify(m pb.Message) {
> 	assertTrue(r.state != StateLeader, "leaders should locally fortify without sending a message")
> ```

Can this even be `r.state == StateFollower`?
___
*[`pkg/raft/rafttest/interaction_env_handler.go` line 208 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48ft6z9s8wVo63c1mj:-O48ft6z9s8wVo63c1mk:b-4wk8ej) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler.go#L208)):*
> ```Go
> 	case "bump-epoch":
> 		// Bumps the epoch of a store. As a result, the store stops seeking support
> 		// from remote stores at the prior epoch. It instead (successfully) seeks
> ```

"from all remote stores" to better distinguish this from the next command.
___
*[`pkg/raft/rafttest/interaction_env_handler.go` line 217 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48gB7XE5qZ6wzcNSjG:-O48gB7Y-tb5fpizK1lR:b76tm5h) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler.go#L217)):*
> ```Go
> 	case "bump-support-for":
> 		// Bumps the support for another store (s_remote), resulting in s_remote
> 		// in s_remote bumping its epoch and using it to seek support from s_local
> ```

"in s_remote in s_remote"
___
*[`pkg/raft/rafttest/interaction_env_handler.go` line 221 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48hYhKA207x3v2OHGq:-O48hYhKA207x3v2OHGr:b-3i3de7) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler.go#L221)):*
> ```Go
> 		//
> 		// Example:
> 		// bump-support-for 1 2
> ```

In addition to the example, it would be nice to include the grammar for each of these commands, because without that it it's difficult to understand what each argument represents. Something like we have above with:
```
// propose-conf-change node_id [v1=<bool>] [transition=<string>]
// Arguments are:
// ...
```
___
*[`pkg/raft/rafttest/interaction_env_handler.go` line 223 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48ggpvEkqZbCg5gNSO:-O48ggpvEkqZbCg5gNSP:b8gb8dw) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler.go#L223)):*
> by store

Is this "by store" terminology coming from the design doc? We've mostly scrapped that in favor of "support from" and "support for", which we decided was less confusing. If so, let's switch this to say "from store". 
___
*[`pkg/raft/rafttest/interaction_env_handler.go` line 223 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48m1KE8w6gHn7YKo3i:-O48m1KE8w6gHn7YKo3j:b-9o7cqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler.go#L223)):*
> ```Go
> 		// bump-support-for 1 2
> 		// Explanation:
> 		// 2 (by store) stops supporting 1's (for store) current epoch. 1 bumps its
> ```

For this command and the next, it feels unnatural to me (and caught me off guard when reading the datadriven test) to have the second argument be the "receiver" and the first argument be the "target". Here we're saying that 2 makes a local decision about 1 in 2's `StoreLiveness.SupportFor` map, and yet 1 is mentioned first.
___
*[`pkg/raft/rafttest/interaction_env_handler.go` line 230 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48hHEF60dsjZkPDgac:-O48hHEF60dsjZkPDgad:bwx6as3) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler.go#L230)):*
> ```Go
> 	case "withdraw-support-for":
> 		// Withdraws support for a remote store's epoch, leaving the epoch
> 		// unchanged.
> ```

What does "leaving the epoch unchanged" mean? When a store withdraws support for another store in store liveness, it advances its epoch and sets the expiration to zero.
___
*[`pkg/raft/rafttest/interaction_env_handler.go` line 234 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48maMo-TOigh-i2wEF:-O48maMpCOnXJqmqv9Mj:bw8kczw) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler.go#L234)):*
> ```Go
> 		// Note that after invoking "withdraw-support-for", a test may establish
> 		// support for the remote store at a higher epoch by calling
> 		// "bump-support-for".
> ```

I struggled to understand why we need "bump-support-for" and "withdraw-support-for". We only have a single notion at the store liveness level (withdraw), so it would be nice to mirror that behavior to avoid surprises. I think this comment hints at how we could make these commands more closely follow the real store liveness. Instead of a "bump-support-for" and "withdraw-support-for", we could have a "withdraw-support-for" and a "grant-support-for". To "bump" support for, the supporter would withdraw support and then grant it again at the later epoch. This would match the direction that information flows in store liveness, where a "requestor" store bumps its epoch in response to a "supporter" withdrawing a previous epoch.
___
*[`pkg/raft/rafttest/interaction_env_handler.go` line 239 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48hA1oB0BGEHEvpfIe:-O48hA1oB0BGEHEvpfIf:b-4bwil0) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler.go#L239)):*
> ```Go
> 		// withdraw-support-for 1 2
> 		// Explanation:
> 		// 2 (by store) withdraws support for 1's (for store) current epoch. All
> ```

Same thing about "by store".
___
*[`pkg/raft/rafttest/interaction_env_handler_store_liveness.go` line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48i7Nb7zzBkpk6z6fC:-O48i7NeCyOspSUFrTWy:bu3n8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler_store_liveness.go#L32)):*
> ```Go
> // livenessFabric is a global view of the store liveness state.
> type livenessFabric struct {
> 	// sate is a 2D array, where state[i][j] represents store i's support for
> ```

`s/sate/state/`
___
*[`pkg/raft/rafttest/interaction_env_handler_store_liveness.go` line 40 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48ilTvBtfI-92HI0t7:-O48ilTyDWp-wyLmZtBV:btsmg7z) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler_store_liveness.go#L40)):*
> ```Go
> func newLivenessFabric() *livenessFabric {
> 	// Stores are 1-indexed. Pad our state with an empty, unusued throughout,
> 	// liveness entry.
> ```

Could you mention that this is helpful for the logic in `addNode`?
___
*[`pkg/raft/rafttest/interaction_env_handler_store_liveness.go` line 56 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48irX59h_9WOqyPIjo:-O48irX59h_9WOqyPIjp:b5p3z8y) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler_store_liveness.go#L56)):*
> ```Go
> 	// For every store that already exists, add another column for the node we're
> 	// adding.
> 	for i := 0; i < len(l.state); i++ {
> ```

nit: `for i := range l.state { ... }` / `for j := range l.state[i] { ... }` here and throughout this file?
___
*[`pkg/raft/rafttest/interaction_env_handler_store_liveness.go` line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48jC0XD-1eZNZOqK2C:-O48jC0XD-1eZNZOqK2D:b-ikqd76) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler_store_liveness.go#L74)):*
> ```Go
> 	// Finally, start the epoch for the node we've just added from 1. It supports
> 	// itself at this epoch.
> 	l.state[len(l.state)-1][len(l.state)-1].epoch = 1
> ```

This would be cleaner if you just assigned a `livenessEntry` struct. Or better yet, pull out a `var initLivenessEntry = livenessEntry{...}` value that you can use throughout this file to avoid duplication.
___
*[`pkg/raft/rafttest/interaction_env_handler_store_liveness.go` line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48k7Wo1NyOs0Gt8TZj:-O48k7Wq0YwgH56QMfNc:bozccnm) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler_store_liveness.go#L131)):*
> ```Go
> func (s *storeLiveness) SupportFrom(id pb.PeerID) (pb.Epoch, hlc.Timestamp, bool) {
> 	entry := s.livenessFabric.state[id][s.nodeID]
> 	// TODO(arul): we may need to inject timestamps in here as well.
> ```

Consider at least returning an empty timestamp if `!entry.isSupported`.
___
*[`pkg/raft/rafttest/interaction_env_handler_store_liveness.go` line 150 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48kLQc-yJZ03eBzA8i:-O48kLQc-yJZ03eBzA8j:bpl5ge8) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler_store_liveness.go#L150)):*
> ```Go
> func (env *InteractionEnv) handleBumpEpoch(t *testing.T, d datadriven.TestData) error {
> 	n := firstAsInt(t, d)
> 	for i := 1; i < len(env.Fabric.state[n]); i++ {
> ```

Should we just start at 0? It's not clear to me throughout this file in which cases we're ignoring the zero-valued node and in which cases we're treating it like any other node. We should stick to one or the other.
___
*[`pkg/raft/rafttest/interaction_env_handler_store_liveness.go` line 163 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48klj81VbR1WqJHZAa:-O48klj81VbR1WqJHZAb:b-zhs41j) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler_store_liveness.go#L163)):*
> ```Go
> func (env *InteractionEnv) handleBumpSupportFor(t *testing.T, d datadriven.TestData) error {
> 	forStore := nthAsInt(t, d, 0)
> 	byStore := nthAsInt(t, d, 1)
> ```

Same comment about "by" being replaced by "from" to align with the terminology in the store liveness layer.
___
*[`pkg/raft/rafttest/interaction_env_handler_store_liveness.go` line 166 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48kwyqCc_zQ7s7RicZ:-O48kwytB7dL9EfzAR2e:b-up4an7) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler_store_liveness.go#L166)):*
> ```Go
> 	// First, bump the forStore's epoch.
> 	env.Fabric.state[forStore][forStore].epoch++
> 	env.Fabric.state[byStore][forStore] = env.Fabric.state[forStore][forStore]
> ```

Is this just grabbing the new epoch field value? If so, maybe make that explicit.
___
*[`pkg/raft/rafttest/interaction_env_handler_store_liveness.go` line 178 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48lB8Z-OJ4le9FkMvS:-O48lB8Z-OJ4le9FkMvT:b-7xfnwp) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler_store_liveness.go#L178)):*
> ```Go
> 	forStore := nthAsInt(t, d, 0)
> 	byStore := nthAsInt(t, d, 1)
> 	env.Fabric.state[byStore][forStore].isSupported = false
> ```

But we're still leaving `env.Fabric.state[forStore][byStore].isSupported` set to true?
___
*[`pkg/raft/testdata/store_liveness_basic.txt` line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48oni_7cnyv59uHZJU:-O48onicEfJggjgducvz:br45bd3) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/testdata/store_liveness_basic.txt#L41)):*
> ```help
> 3 2 1 1
> 
> # Withdraw support for 2 by 3.
> ```

nit: a few other cases of "by" which would be better as "from".


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
Thanks for the look, should be good for another. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-:-O4C9f05EB90sl8Agqrz:b-hljs4q)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/kv/kvserver/replica_store_liveness.go` line 48 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48eOsaAXKmdbz011zG:-O4Be7jO-S0aH_ketU7z:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/kv/kvserver/replica_store_liveness.go#L48)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we want to keep this indefinitely, or add a TODO to remove it once `storeLiveness` is configured in the `Store` constructor?
</blockquote></details>

Done.
___
*[`pkg/raft/raft.go` line 521 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48f547CJIxVhrtL8MT:-O4BglSF6LM2Wh9lO7e4:byz45ny) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/raft.go#L521)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Reminder that we're going to need to add `MsgFortifyResp` to this list. We may want to do so now, to avoid the data-driven churn.
</blockquote></details>

Done, and updated the comment. I didn't (out of laziness) pull it out into its own commit (maybe I should've) -- let me know if you want me to.
___
*[`pkg/raft/raft.go` line 1713 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48fH1JDlqy6qJNAk4C:-O4BhRKACWQFyU0-MygV:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/raft.go#L1713)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We haven't quite decided how we want to use the `r.electionElapsed` field with fortification, but we may end up wanting to reset it to 0 here, for parity with the other cases.
</blockquote></details>

Done. 
___
*[`pkg/raft/raft.go` line 1881 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48en2g9rBPkM8st_1F:-O4BhBOA0cebh5whUibY:b9mz5n4) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/raft.go#L1881)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Can this even be `r.state == StateFollower`?
</blockquote></details>

Yeah, it can be. `handleFortify` is called from `stepCandidate`, but we call `becomeFollower` before calling into here.
___
*[`pkg/raft/rafttest/interaction_env_handler.go` line 221 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48hYhKA207x3v2OHGq:-O4C1QweDKZd48-NjoRG:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler.go#L221)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

In addition to the example, it would be nice to include the grammar for each of these commands, because without that it it's difficult to understand what each argument represents. Something like we have above with:
```
// propose-conf-change node_id [v1=<bool>] [transition=<string>]
// Arguments are:
// ...
```
</blockquote></details>

Done. 
___
*[`pkg/raft/rafttest/interaction_env_handler.go` line 223 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48m1KE8w6gHn7YKo3i:-O4BwANeDY65XSmKyasa:btnpjr7) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler.go#L223)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

For this command and the next, it feels unnatural to me (and caught me off guard when reading the datadriven test) to have the second argument be the "receiver" and the first argument be the "target". Here we're saying that 2 makes a local decision about 1 in 2's `StoreLiveness.SupportFor` map, and yet 1 is mentioned first.
</blockquote></details>

Heh I've been reading these as "bump-support-for <for_id> [from] <from_id>". But now I'm second guessing myself after your comment -- I'll just switch it.

----

EDIT: Coming back to this, I'm gonna drop the "for" suffix so that I don't confuse myself.
___
*[`pkg/raft/rafttest/interaction_env_handler.go` line 223 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48ggpvEkqZbCg5gNSO:-O4C16fa6xxbpwY7Fuq9:bkmdhfc) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler.go#L223)):*
> Is this "by store" terminology coming from the design doc?

Yeah. 

> We've mostly scrapped that in favor of "support from" and "support for", which we decided was less confusing. If so, let's switch this to say "from store".

Sgtm. Switched; I think I got everything, but let me know if something's still missed. 
___
*[`pkg/raft/rafttest/interaction_env_handler.go` line 230 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48hHEF60dsjZkPDgac:-O4C3fkC0muYjIzdbuUM:bq2ixuq) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler.go#L230)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

What does "leaving the epoch unchanged" mean? When a store withdraws support for another store in store liveness, it advances its epoch and sets the expiration to zero.
</blockquote></details>

Changed things around to match store liveness. I've also extended `store_liveness_basic` to include an edge case that falls out of making this change. 
___
*[`pkg/raft/rafttest/interaction_env_handler.go` line 234 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48maMo-TOigh-i2wEF:-O4C3kFu7w0w0uicL-SW:b-yuu0wl) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler.go#L234)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I struggled to understand why we need "bump-support-for" and "withdraw-support-for". We only have a single notion at the store liveness level (withdraw), so it would be nice to mirror that behavior to avoid surprises. I think this comment hints at how we could make these commands more closely follow the real store liveness. Instead of a "bump-support-for" and "withdraw-support-for", we could have a "withdraw-support-for" and a "grant-support-for". To "bump" support for, the supporter would withdraw support and then grant it again at the later epoch. This would match the direction that information flows in store liveness, where a "requestor" store bumps its epoch in response to a "supporter" withdrawing a previous epoch.
</blockquote></details>

Made some changes. Lemme know if this was what you had in mind.
___
*[`pkg/raft/rafttest/interaction_env_handler_store_liveness.go` line 40 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48ilTvBtfI-92HI0t7:-O4C5lpl5CqMtEmNN8xb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler_store_liveness.go#L40)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Could you mention that this is helpful for the logic in `addNode`?
</blockquote></details>

Done.
___
*[`pkg/raft/rafttest/interaction_env_handler_store_liveness.go` line 74 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48jC0XD-1eZNZOqK2C:-O4C7TItAZqyfbprpdY_:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler_store_liveness.go#L74)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This would be cleaner if you just assigned a `livenessEntry` struct. Or better yet, pull out a `var initLivenessEntry = livenessEntry{...}` value that you can use throughout this file to avoid duplication.
</blockquote></details>

Done. 
___
*[`pkg/raft/rafttest/interaction_env_handler_store_liveness.go` line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48k7Wo1NyOs0Gt8TZj:-O4C8p3F1sCit9rHk6ka:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler_store_liveness.go#L131)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Consider at least returning an empty timestamp if `!entry.isSupported`.
</blockquote></details>

Done.
___
*[`pkg/raft/rafttest/interaction_env_handler_store_liveness.go` line 150 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48kLQc-yJZ03eBzA8i:-O4C9JY_5hqBpcP6gvEq:bswpsyg) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler_store_liveness.go#L150)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should we just start at 0? It's not clear to me throughout this file in which cases we're ignoring the zero-valued node and in which cases we're treating it like any other node. We should stick to one or the other.
</blockquote></details>

Switched to starting at 0. 
___
*[`pkg/raft/rafttest/interaction_env_handler_store_liveness.go` line 166 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48kwyqCc_zQ7s7RicZ:-O4BsxMMC4V5A44drnqO:bydmaa5) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler_store_liveness.go#L166)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this just grabbing the new epoch field value? If so, maybe make that explicit.
</blockquote></details>

Yeah. This was meant to be: 

```golang
	env.Fabric.state[fromStore][forStore].epoch = env.Fabric.state[forStore][forStore].epoch
```

Probably an autocomplete misstep; that was a weird line of code to read back.
___
*[`pkg/raft/rafttest/interaction_env_handler_store_liveness.go` line 178 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48lB8Z-OJ4le9FkMvS:-O4Btr7p4ROrKYIl0iYp:b1lhedh) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler_store_liveness.go#L178)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

But we're still leaving `env.Fabric.state[forStore][byStore].isSupported` set to true?
</blockquote></details>

Yeah -- I was imagining if you wanted to create a "bi-directional withdraw" you'd do so using 2 datadriven commands. This way, there's more flexibility. 

Thoughts?


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 36 of 36 files at r3, 36 of 36 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128845)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/raft/raft.go` line 521 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48f547CJIxVhrtL8MT:-O4CD4CFBfMRoRjPNaSd:bqptlu1) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/raft.go#L521)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Done, and updated the comment. I didn't (out of laziness) pull it out into its own commit (maybe I should've) -- let me know if you want me to.
</blockquote></details>

No need, this commit is fine.
___
*[`pkg/raft/raft.go` line 530 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O4CDBBb05CKNuyU03Yd:-O4CDBBb05CKNuyU03Ye:b-x7ydws) ([raw file](https://github.com/cockroachdb/cockroach/blob/57e7247aeb24f8e0a7e55f9670f4a3880d07ddf9/pkg/raft/raft.go#L530)):*
> ```Go
> 		// For most message types, this is not an issue. However, response messages
> 		// that relate to "voting" on either leader election or log appends or
> 		// messages that relate to leader fortification require durability before
> ```

nit: you can shorten this to just "either leader election or log appends or leader fortification". No need for "that related to" twice.
___
*[`pkg/raft/rafttest/interaction_env_handler.go` line 234 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48maMo-TOigh-i2wEF:-O4CE8LOE7Xxz1kA4fRM:b-8jyzc6) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler.go#L234)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Made some changes. Lemme know if this was what you had in mind.
</blockquote></details>

Looks good to me. Thanks for making the change, it's easier to understand.
___
*[`pkg/raft/rafttest/interaction_env_handler.go` line 241 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O4CDz5C-gOKEoUeFfpT:-O4CDz5C-gOKEoUeFfpU:bnw6yk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/57e7247aeb24f8e0a7e55f9670f4a3880d07ddf9/pkg/raft/rafttest/interaction_env_handler.go#L241)):*
> ```Go
> 		// its epoch and using it to seek support from from_store at this new epoch.
> 		//
> 		// Note that From_store should not be supporting for_store already; if it
> ```

nit: s/From_store/from_store/
___
*[`pkg/raft/rafttest/interaction_env_handler.go` line 246 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O4CE1z9-nqKGAIpTQtX:-O4CE1z9-nqKGAIpTQtY:bf24exg) ([raw file](https://github.com/cockroachdb/cockroach/blob/57e7247aeb24f8e0a7e55f9670f4a3880d07ddf9/pkg/raft/rafttest/interaction_env_handler.go#L246)):*
> "withdrawing support"

These two lines need an update.
___
*[`pkg/raft/rafttest/interaction_env_handler_store_liveness.go` line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48k7Wo1NyOs0Gt8TZj:-O4CEi5s-0Ri4Z7YKPOW:b-sbht3a) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler_store_liveness.go#L131)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

`hlc.MinTimestamp != hlc.Timestamp{}`. We typically use the former to mean the "minimum valid timestamp", while the latter means an "invalid timestamp". So we probably want `hlc.Timestamp{}`.
___
*[`pkg/raft/rafttest/interaction_env_handler_store_liveness.go` line 178 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48lB8Z-OJ4le9FkMvS:-O4CFNDR52UEJAG1BR-s:b-ljp08p) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler_store_liveness.go#L178)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Yeah -- I was imagining if you wanted to create a "bi-directional withdraw" you'd do so using 2 datadriven commands. This way, there's more flexibility. 

Thoughts?
</blockquote></details>

Between any two stores (`s1` and `s2`), there are 4 `isSupported` bools.
1. `env.Fabric.state[s1][s1].isSupported`
2. `env.Fabric.state[s1][s2].isSupported`
3. `env.Fabric.state[s2][s1].isSupported`
4. `env.Fabric.state[s2][s2].isSupported`

Are all of these meaningful?
___
*[`pkg/raft/rafttest/interaction_env_handler_store_liveness.go` line 67 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O4CESnV1DE92WqfQTUK:-O4CESnV1DE92WqfQTUL:bl9nvw5) ([raw file](https://github.com/cockroachdb/cockroach/blob/57e7247aeb24f8e0a7e55f9670f4a3880d07ddf9/pkg/raft/rafttest/interaction_env_handler_store_liveness.go#L67)):*
> ```Go
> 	newNodeState := make([]livenessEntry, len(l.state)+1) // 1-indexed stores
> 	for i := range l.state {
> 		newNodeState[i] = livenessEntry{
> ```

Should this just be `newNodeState[i] = l.state[i][i]`? Are there ever cases where `l.state[i][i].isSupported == false`? If so, do we want other nodes to provide support in those cases?
___
*[`pkg/raft/testdata/async_storage_writes.txt` line 117 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O4CFi7y7RbPsnJb6big:-O4CFi7y7RbPsnJb6bih:badgwwc) ([raw file](https://github.com/cockroachdb/cockroach/blob/57e7247aeb24f8e0a7e55f9670f4a3880d07ddf9/pkg/raft/testdata/async_storage_writes.txt#L117)):*
> ```help
>   1/11 EntryNormal ""
>   Messages:
>   2->AppendThread MsgStorageAppend Term:1 Log:1/11 Commit:10 Vote:1 Lead:1 Entries:[1/11 EntryNormal ""] Responses:[
> ```

Once we start assigning LeadEpoch to `MsgStorageAppend` in `newStorageAppendMsg`, we'll see it in this output. Was your plan to do that in a follow up PR?


<!-- Sent from Reviewable.io -->

arulajmani committed:
raft: include LeadEpoch in MsgStorageAppend

Previously, we weren't handling HardState changes correctly in
MsgStorageAppend. Moreover, we weren't including LeadEpoch on the
message, so it would never get persisted as intended. This patch
fixes this issue.

Epic: none

Release note: None

arulajmani reviewed (commented):
RFAL! 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-:-O4Cpm2Q6NK0kIjZsdTS:by76qkj)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten)

___
*[`pkg/raft/rafttest/interaction_env_handler_store_liveness.go` line 131 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48k7Wo1NyOs0Gt8TZj:-O4CPTIW01hpEWxFJfv8:b-ji8rzq) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler_store_liveness.go#L131)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`hlc.MinTimestamp != hlc.Timestamp{}`. We typically use the former to mean the "minimum valid timestamp", while the latter means an "invalid timestamp". So we probably want `hlc.Timestamp{}`.
</blockquote></details>

Ack. Done. 
___
*[`pkg/raft/rafttest/interaction_env_handler_store_liveness.go` line 178 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O48lB8Z-OJ4le9FkMvS:-O4CnhBOEl68Dqm_68X_:bru50pv) ([raw file](https://github.com/cockroachdb/cockroach/blob/53bbbc583ccdc7bc244754b863852b492bf54001/pkg/raft/rafttest/interaction_env_handler_store_liveness.go#L178)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Between any two stores (`s1` and `s2`), there are 4 `isSupported` bools.
1. `env.Fabric.state[s1][s1].isSupported`
2. `env.Fabric.state[s1][s2].isSupported`
3. `env.Fabric.state[s2][s1].isSupported`
4. `env.Fabric.state[s2][s2].isSupported`

Are all of these meaningful?
</blockquote></details>

We spoke about this offline and decided to keep things as is.
___
*[`pkg/raft/rafttest/interaction_env_handler_store_liveness.go` line 67 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O4CESnV1DE92WqfQTUK:-O4CN4J05ehQw7JSQhvR:baii9t3) ([raw file](https://github.com/cockroachdb/cockroach/blob/57e7247aeb24f8e0a7e55f9670f4a3880d07ddf9/pkg/raft/rafttest/interaction_env_handler_store_liveness.go#L67)):*
> Are there ever cases where `l.state[i][i].isSupported == false`?

Yeah, in cases where a node withdraws support from itself. 

> If so, do we want other nodes to provide support in those cases?

Other nodes should be able to provide support even if the node isn't supporting itself. It's not clear whether a new node added should automatically provide support or not in this case, but I think it's easier to say a newly added node supports all other nodes until explicitly told not to.
___
*[`pkg/raft/testdata/async_storage_writes.txt` line 117 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-O4CFi7y7RbPsnJb6big:-O4Cnrck3Ep63WxRik3A:b-xbzqt0) ([raw file](https://github.com/cockroachdb/cockroach/blob/57e7247aeb24f8e0a7e55f9670f4a3880d07ddf9/pkg/raft/testdata/async_storage_writes.txt#L117)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Once we start assigning LeadEpoch to `MsgStorageAppend` in `newStorageAppendMsg`, we'll see it in this output. Was your plan to do that in a follow up PR?
</blockquote></details>

Spoke about this offline, this was an oversight; added r3 to fix this.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r5, 4 of 4 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/128845#-:-O4DRifvAZ9CKu4CK8KB:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @arulajmani)



<!-- Sent from Reviewable.io -->

arulajmani commented:
Thanks for the speedy reviews on this one! 

bors r=nvanbenschoten

arulajmani commented:
I'd missed updating some Raft copyright headers; fixed. 

bors r=nvanbenschoten
